plugins/net_plugin/include/eosio/net_plugin/protocol.hpp

```diff
@@ -137,22 +137,27 @@ namespace eosio {
    };
 
    struct gossip_bp_peers_message {
+      struct bp_peer_info_v1 {
+         std::string               server_endpoint;      // externally available address to connect to
+         std::string               outbound_ip_address;  // outbound ip address for firewall
+         block_timestamp_type      expiration;           // head block to remove bp_peer
+      };
+      // bp_peer_info_v2 can derive from bp_peer_info_v1, so old peers can still unpack bp_peer_info_v1 from bp_peer::bp_peer_info
       struct bp_peer {
+         unsigned_int              version = 1;
          eosio::name               producer_name;
-         std::string               server_endpoint;         // externally available address to connect to
-         std::string               outbound_ip_address;     // empty if equal to server_endpoint ip address
-         block_timestamp_type      expiration;              // head block to remove bp_peer
+         std::vector<char>         bp_peer_info;         // serialized bp_peer_info
 
          digest_type digest() const;
-
-         bool operator==(const bp_peer&) const = default;
-         bool operator<(const bp_peer& rhs) const {
-            // [producer_name, server_endpoint] is unique
-            return std::tie(producer_name, server_endpoint) < std::tie(rhs.producer_name, rhs.server_endpoint);
-         }
       };
       struct signed_bp_peer : bp_peer {
-         signature_type            sig;
+         signature_type  sig; // signature over bp_peer
+
+         std::optional<bp_peer_info_v1> cached_bp_peer_info; // not serialized
```

Comment:
Why not follow the pattern in SHiP and TraceAPI that use `std::variant` to represent versions?

Reply:
We can do that when we have another version. No reason for a variant now.